File: src/main/java/net/minestom/server/entity/pathfinding/PFPathingEntity.java
package net.minestom.server.entity.pathfinding;

import com.extollit.gaming.ai.path.model.Gravitation;
import com.extollit.gaming.ai.path.model.IPathingEntity;
import com.extollit.gaming.ai.path.model.Passibility;
import com.extollit.linalg.immutable.Vec3d;
import net.minestom.server.attribute.Attribute;
import net.minestom.server.attribute.AttributeImpl;
import net.minestom.server.attribute.VanillaAttribute;
import net.minestom.server.coordinate.Point;
import net.minestom.server.coordinate.Vec;
import net.minestom.server.entity.Entity;
import net.minestom.server.entity.LivingEntity;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

@ApiStatus.Internal
public final class PFPathingEntity implements IPathingEntity {
    private final Navigator navigator;
    private final Entity entity;

    private float searchRange;

    // Capacities
    private boolean fireResistant;
    private boolean cautious;
    private boolean climber;
    private boolean swimmer;
    private boolean aquatic;
    private boolean avian;
    private boolean aquaphobic;
    private boolean avoidsDoorways;
    private boolean opensDoors;

    public PFPathingEntity(Navigator navigator) {
        this.navigator = navigator;
        this.entity = navigator.getEntity();

        this.searchRange = getAttributeValue(VanillaAttribute.GENERIC_FOLLOW_RANGE);
    }

    @Override
    public int age() {
        return (int) entity.getAliveTicks();
    }

    @Override
    public boolean bound() {
        return entity.hasVelocity();
    }

    @Override
    public float searchRange() {
        return searchRange;
    }

    /**
     * Changes the search range of the entity
     *
     * @param searchRange the new entity's search range
     */
    public void setSearchRange(float searchRange) {
        this.searchRange = searchRange;
    }

    public boolean isFireResistant() {
        return fireResistant;
    }

    public void setFireResistant(boolean fireResistant) {
        this.fireResistant = fireResistant;
    }

    public boolean isCautious() {
        return cautious;
    }

    public void setCautious(boolean cautious) {
        this.cautious = cautious;
    }

    public boolean isClimber() {
        return climber;
    }

    public void setClimber(boolean climber) {
        this.climber = climber;
    }

    public boolean isSwimmer() {
        return swimmer;
    }

    public void setSwimmer(boolean swimmer) {
        this.swimmer = swimmer;
    }

    public boolean isAquatic() {
        return aquatic;
    }

    public void setAquatic(boolean aquatic) {
        this.aquatic = aquatic;
    }

    public boolean isAvian() {
        return avian;
    }

    public void setAvian(boolean avian) {
        this.avian = avian;
    }

    public boolean isAquaphobic() {
        return aquaphobic;
    }

    public void setAquaphobic(boolean aquaphobic) {
        this.aquaphobic = aquaphobic;
    }

    public boolean isAvoidsDoorways() {
        return avoidsDoorways;
    }

    public void setAvoidsDoorways(boolean avoidsDoorways) {
        this.avoidsDoorways = avoidsDoorways;
    }

    public boolean isOpensDoors() {
        return opensDoors;
    }

    public void setOpensDoors(boolean opensDoors) {
        this.opensDoors = opensDoors;
    }

    @Override
    public Capabilities capabilities() {
        return new Capabilities() {
            @Override
            public float speed() {
                return getAttributeValue(VanillaAttribute.GENERIC_MOVEMENT_SPEED);
            }

            @Override
            public boolean fireResistant() {
                return fireResistant;
            }

            @Override
            public boolean cautious() {
                return cautious;
            }

            @Override
            public boolean climber() {
                return climber;
            }

            @Override
            public boolean swimmer() {
                return swimmer;
            }

            @Override
            public boolean aquatic() {
                return aquatic;
            }

            @Override
            public boolean avian() {
                return avian;
            }

            @Override
            public boolean aquaphobic() {
                return aquaphobic;
            }

            @Override
            public boolean avoidsDoorways() {
                return avoidsDoorways;
            }

            @Override
            public boolean opensDoors() {
                return opensDoors;
            }
        };
    }

    @Override
    public void moveTo(Vec3d position, Passibility passibility, Gravitation gravitation) {
        final Point targetPosition = new Vec(position.x, position.y, position.z);
        this.navigator.moveTowards(targetPosition, getAttributeValue(VanillaAttribute.GENERIC_MOVEMENT_SPEED));
        final double entityY = entity.getPosition().y() + 0.00001D; // After any negative y movement, entities will always be extremely
                                                                    // slightly below floor level. This +0.00001D is here to offset this
                                                                    // error and stop the entity from permanently jumping.

        if (entityY < targetPosition.y()) {
            this.navigator.jump(1);
        }
    }

    @Override
    public Vec3d coordinates() {
        final var position = entity.getPosition();
        return new Vec3d(position.x(), position.y(), position.z());
    }

    @Override
    public float width() {
        return (float) entity.getBoundingBox().width();
    }

    @Override
    public float height() {
        return (float) entity.getBoundingBox().height();
    }

    private float getAttributeValue(@NotNull Attribute attribute) {
        if (entity instanceof LivingEntity) {
            return ((LivingEntity) entity).getAttributeValue(attribute);
        }
        return 0f;
    }
}

File: src/main/java/net/minestom/server/entity/pathfinding/PFPathingEntity.kt
package net.minestom.server.entity.pathfinding

import net.minestom.server.utils.position.PositionUtils.getLookYaw
import net.minestom.server.utils.position.PositionUtils.getLookPitch
import net.minestom.server.utils.chunk.ChunkUtils.isLoaded
import com.extollit.gaming.ai.path.model.IBlockDescription
import com.extollit.gaming.ai.path.model.IBlockObject
import space.vectrix.flare.fastutil.Short2ObjectSyncMap
import net.minestom.server.entity.pathfinding.PFBlock
import net.minestom.server.entity.pathfinding.PFPathingEntity
import com.extollit.gaming.ai.path.HydrazinePathFinder
import net.minestom.server.collision.PhysicsResult
import net.minestom.server.collision.CollisionUtils
import net.minestom.server.coordinate.Vec
import net.minestom.server.instance.WorldBorder
import net.minestom.server.utils.chunk.ChunkUtils
import com.extollit.gaming.ai.path.PathOptions
import com.extollit.gaming.ai.path.model.IPath
import net.minestom.server.entity.LivingEntity
import net.minestom.server.entity.pathfinding.PFInstanceSpace
import com.extollit.gaming.ai.path.model.IColumnarSpace
import com.extollit.gaming.ai.path.model.ColumnarOcclusionFieldList
import com.extollit.gaming.ai.path.model.IInstanceSpace
import net.minestom.server.entity.pathfinding.PFColumnarSpace
import java.util.concurrent.ConcurrentHashMap
import com.extollit.gaming.ai.path.model.IPathingEntity
import com.extollit.gaming.ai.path.model.Passibility
import com.extollit.gaming.ai.path.model.Gravitation
import com.extollit.linalg.immutable.Vec3d
import net.minestom.server.attribute.Attribute
import net.minestom.server.coordinate.Point
import net.minestom.server.entity.Entity
import org.jetbrains.annotations.ApiStatus

@ApiStatus.Internal
class PFPathingEntity(private val navigator: Navigator) : IPathingEntity {
    private val entity: Entity
    private var searchRange: Float

    // Capacities
    var isFireResistant = false
    var isCautious = false
    var isClimber = false
    var isSwimmer = false
    var isAquatic = false
    var isAvian = false
    var isAquaphobic = false
    var isAvoidsDoorways = false
    var isOpensDoors = false

    init {
        entity = navigator.entity
        searchRange = getAttributeValue(Attribute.FOLLOW_RANGE)
    }

    override fun age(): Int {
        return entity.aliveTicks.toInt()
    }

    override fun bound(): Boolean {
        return entity.hasVelocity()
    }

    override fun searchRange(): Float {
        return searchRange
    }

    /**
     * Changes the search range of the entity
     *
     * @param searchRange the new entity's search range
     */
    fun setSearchRange(searchRange: Float) {
        this.searchRange = searchRange
    }

    override fun capabilities(): IPathingEntity.Capabilities {
        return object : IPathingEntity.Capabilities {
            override fun speed(): Float {
                return getAttributeValue(Attribute.MOVEMENT_SPEED)
            }

            override fun fireResistant(): Boolean {
                return isFireResistant
            }

            override fun cautious(): Boolean {
                return isCautious
            }

            override fun climber(): Boolean {
                return isClimber
            }

            override fun swimmer(): Boolean {
                return isSwimmer
            }

            override fun aquatic(): Boolean {
                return isAquatic
            }

            override fun avian(): Boolean {
                return isAvian
            }

            override fun aquaphobic(): Boolean {
                return isAquaphobic
            }

            override fun avoidsDoorways(): Boolean {
                return isAvoidsDoorways
            }

            override fun opensDoors(): Boolean {
                return isOpensDoors
            }
        }
    }

    override fun moveTo(position: Vec3d, passibility: Passibility, gravitation: Gravitation) {
        val targetPosition: Point = Vec(position.x, position.y, position.z)
        navigator.moveTowards(targetPosition, getAttributeValue(Attribute.MOVEMENT_SPEED).toDouble())
        val entityY = entity.position.y()
        if (entityY < targetPosition.y()) {
            navigator.jump(1f)
        }
    }

    override fun coordinates(): Vec3d {
        val position = entity.position
        return Vec3d(position.x(), position.y(), position.z())
    }

    override fun width(): Float {
        return entity.boundingBox.width().toFloat()
    }

    override fun height(): Float {
        return entity.boundingBox.height().toFloat()
    }

    private fun getAttributeValue(attribute: Attribute): Float {
        return if (entity is LivingEntity) {
            entity.getAttributeValue(attribute)
        } else 0f
    }
}